packages/pg-connection-string/index.js

```diff
@@ -87,15 +87,26 @@ function parse(str) {
       break
     }
     case 'prefer':
-    case 'require':
-    case 'verify-ca':
-    case 'verify-full': {
-      break
-    }
     case 'no-verify': {
       config.ssl.rejectUnauthorized = false
       break
     }
+    case 'require': {
+      if (config.sslrootcert) {
+        // If a root CA is specified, behavior of `sslmode=require` will be the same as that of `verify-ca`
+        config.ssl.checkServerIdentity = function () {}
+      } else {
+        config.ssl.rejectUnauthorized = false
+      }
+      break
+    }
+    case 'verify-ca': {
+      config.ssl.checkServerIdentity = function () {}
```

Comment:
On second thought, even though libpq has and has documented this same problem, maybe we should enforce that `config.ssl.ca` is truthy here.

```suggestion
      if (!config.ssl.ca) {
        throw new Error('sslmode=verify-ca requires specifying a CA with sslrootcert')
      }
      config.ssl.checkServerIdentity = function () {}
```

The main issue with this would be for anyone who wants to add a `ca` key to the returned `config.ssl` *after* parsing. I don’t think a warning is the way to go (too easy for those who need it to miss or ignore, and annoying for those who don’t need it to suppress). Would adding a whole options parameter to support this (`parse(str, {deferSslCa: true})` – probably a better name option) be too much? (Related issue, merging configuration with a connection string is hard in general: #3327)

Reply:
added the check in https://github.com/brianc/node-postgres/pull/2709/commits/74a314ef959bcddd323c138eaacc1b8212e991ed

Reply:
Is it really needed to enforce `config.ssl.ca` for this sslmode?

If `config.ssl.ca` is not specified, the TLS client will use the CA certificates already available via `tls.getCACertificates('default')`, which can include certificates included in the bundled certificate store, the system's trusted store and any extra certificate configured via the `NODE_EXTRA_CA_CERTS` env variable. [Docs](https://nodejs.org/api/tls.html#tlsgetcacertificatestype)

In our implementation, we use the latter option (`NODE_EXTRA_CA_CERTS`) to configure the CA certificates from our Postgres endpoint and it works fine.

Reply:
TIL

I will unwind this change. It will match the behavior in https://github.com/lib/pq/commit/585ef5d6d502fd43589d48e448b91b449745e369#diff-4f427d2b022907c552328e63f137561f6de92396d7a6e8f6c2ea1bcf0db52654R836-R838

Reply:
@pmalouin This is exactly why I thought a check might be useful :) Anyone can get a valid certificate from a system-trusted CA, simply by using a domain that they actually own. Without `config.ssl.ca` enforced, such a certificate can then be used to MITM any `sslmode=verify-ca` connection, providing essentially the same security level as `sslmode=require`.

Reply:
This behavior seems consistent, well documented and understood by those who want libpq compatibility. If someone has a MITM concern, then they should use `sslmode=verify-full` or specify a more narrow `sslrootcert`.

Also, I am not sure the proposed check will have the intended results. If a user encounters this error, they will almost certainly add `&sslrootcert=/etc/ssl/certs/ca-certificates.crt` to the connection string and move on. In fact, I originally thought specifying the system CA (or a custom CA) was the intention of the check because I did not realize that the TLS client will use the CA certificates already available.

All that being said, I did see the following chart in the JDBC docs which outlines the risks in a way that is easy to understand. I think we should incorporate a similar chart into the README so people are aware of the MITM risks.

![image](https://github.com/user-attachments/assets/68ab63e1-e24f-4b20-a8e1-7240d4059a90)

Reply:
I mentioned that it was documented by libpq in my original comment, but I think @pmalouin’s comment demonstrates that the implications are still not obvious to people. A more descriptive error could be a good idea. I don’t think any amount of documentation is sufficient to replace the check.

Reply:
I appreciate the value of guiding people in a safe direction. The current behavior is is unchanged: `prefer`, `require` and `verify-ca` are all aliases `verify-full` by default.

The proposed changes are behind a flag that explicitly asks for libpq compatibility. If someone is explicitly asking for libpq behavior by adding `sslcompat=libpq` then we should be consistent with libpq's behavior.

Reply:
Makes sense to me too. The check would act as a safe default, steering users towards the safer way to use `verify-ca`: explicitly setting the CA that they expect to be used by their server, instead of opening it up to any system trusted CA by default 👍

Reply:
Also, in the spirit of following libpq semantics more closely, it appears that libpq does not use the system trusted certificates by default (ex. when `sslrootcert` is omitted). So this would suggest to do the same here.

libpq also supports the `sslrootcert=system` [option](https://www.postgresql.org/docs/16/libpq-connect.html#LIBPQ-CONNECT-SSLROOTCERT) that specifies to use the system trusted CAs, but libpq would return an error if you use that in combination with `sslmode=verify-ca`.

Reply:
I think where we are at is:

1. put the check back in place to match libpq semantics
2. improve the error message to warn folks about mitm attacks
3. make the sslmode documentation tradeoffs more clear, especially in libpq mode

I will send an updated PR soon :tm:

Reply:
I believe https://github.com/brianc/node-postgres/pull/2709/commits/165b5c6e05dadbbea9175a312ccdb4cbaf28c500 resolves the raised concerns. Please let me know if there is still more to discuss.